o1vm/src/mips/tests.rs

```diff
@@ -323,4 +335,200 @@ mod folding {
+        //     .iter()
+        //     .for_each(|constraint| println!("Degree: {:?}", constraint.degree(1, 0)));
+        // Selecting the first constraint for testing
+        // let constraints = vec![constraints.first().unwrap().clone()];
+
+        let witness_one = make_random_witness_for_addiu(domain_size, &mut rng);
+        let witness_two = make_random_witness_for_addiu(domain_size, &mut rng);
+        // FIXME: run PlonK here to check the it is satisfied.
+
+        // Now, we will fold and use the folding scheme to only prove the
+        // aggregation instead of the individual instances
+
+        #[derive(Clone, Debug, PartialEq, Eq, Hash)]
+        struct MIPSFoldingConfig;
+
+        let trace_one: Trace<N_MIPS_REL_COLS, MIPSFoldingConfig> = Trace {
+            domain_size,
+            // FIXME: do not use clone
+            witness: witness_one.clone(),
+            constraints: constraints.clone(),
+            lookups: vec![],
+        };
+
+        impl FoldingConfig for MIPSFoldingConfig {
+            type Column = Column;
+            type Selector = ();
+            type Challenge = Challenge;
+            type Curve = Curve;
+            type Srs = SRS<Curve>;
+            type Instance = FoldingInstance<N_MIPS_REL_COLS, Curve>;
+            type Witness = FoldingWitness<N_MIPS_REL_COLS, Fp>;
+            // The structure must a provable Trace. Here we use a single
+            // instruction trace
+            type Structure = Trace<N_MIPS_REL_COLS, MIPSFoldingConfig>;
+            type Env = FoldingEnvironment<
+                N_MIPS_REL_COLS,
+                MIPSFoldingConfig,
+                Trace<N_MIPS_REL_COLS, MIPSFoldingConfig>,
+            >;
+        }
+
+        let folding_witness_one: FoldingWitness<N_MIPS_REL_COLS, Fp> = {
+            let witness_one = (&witness_one)
+                .into_par_iter()
+                .map(|w| Evaluations::from_vec_and_domain(w.to_vec(), domain))
+                .collect();
+            FoldingWitness {
+                witness: witness_one,
+            }
+        };
+
+        let folding_witness_two: FoldingWitness<N_MIPS_REL_COLS, Fp> = {
+            let witness_two = (&witness_two)
+                .into_par_iter()
+                .map(|w| Evaluations::from_vec_and_domain(w.to_vec(), domain))
+                .collect();
+            FoldingWitness {
+                witness: witness_two,
+            }
+        };
+
+        let folding_instance_one =
+            build_folding_instance(&folding_witness_one, &mut fq_sponge, domain, &srs);
+        let folding_instance_two =
+            build_folding_instance(&folding_witness_two, &mut fq_sponge, domain, &srs);
+
+        let folding_compat_constraints: Vec<FoldingCompatibleExpr<MIPSFoldingConfig>> = constraints
+            .iter()
+            .map(|x| FoldingCompatibleExpr::from(x.clone()))
+            .collect::<Vec<_>>();
+
+        let (folding_scheme, _) = FoldingScheme::<MIPSFoldingConfig>::new(
```

Comment:
I suppose the next step will be to have the same example but using the decomposable scheme instead, right? I am curious to see if it works for ADDIU or not.

Reply:
Yes. Which would be duplicated what you did in Keccak